packages/shared/src/guards.ts

```diff
@@ -388,3 +390,32 @@ export function isTypedNdArray<T extends NumArray>(
 ): ndArr is NdArray<Exclude<T, number[]>> {
   return ndArr.dtype !== 'array';
 }
+
+export function hasCompoundType<S extends Shape>(
```

Comment:
I would maybe move all those new guards/assertions a bit further up, right after the other dtype guards/assertions.